doc/docs/Python_Tutorials/Adjoint_Solver.md

```diff
@@ -2,13 +2,704 @@
+Much of the functionality of the adjoints solver is implemented in
+Python using [autograd](https://github.com/HIPS/autograd) as well as
+[JAX](https://github.com/google/jax).
 
-There are six Jupyter notebooks that demonstrate the main features of the adjoint solver.
+The adjoint solver supports inverse design and [topology
+optimization](https://en.wikipedia.org/wiki/Topology_optimization) by
+providing the functionality to wrap an optimization library around the
+gradient computation. This is demonstrated in several tutorials below.
+
+[TOC]
+
+Broadband Waveguide Mode Converter with Minimum Feature Size
+------------------------------------------------------------
+
+This example demonstrates some of the advanced functionality of the
+adjoint solver including worst-case (minimax) optimization, multiple
+objective functions, and constraints on the minimum line width and
+line spacing. The design problem involves a broadband waveguide mode
+converter with minimum feature size in 2D. This is based on
+[M.F. Schubert et al., ACS Photonics, Vol. 9, pp. 2327-36,
+(2022)](https://doi.org/10.1021/acsphotonics.2c00313).
+
+The mode converter must satisfy two separate design objectives: (1)
+minimize the reflectance $R$ into the fundamental mode of the input
+port ($\mathcal{S}_{11}$) and (2) maximize the transmittance $T$ into
+the second-order mode of the output port ($\mathcal{S}_{21}$). There
+are different ways to define this multi-objective and multi-wavelength
+optimization. The approach taken here is *worst-case* optimization
+whereby we minimize the maximum (the worst case) of $R$ and $1-T$
+across all six wavelengths (in the O-band for
+telecommunications). This is known as *minimax* optimization.
+
+The challenge with minimax optimization is that the $\max$ objective
+function is not everywhere differentiable. This property would seem to
+preclude the use of gradient-based optimization algorithms for this
+problem which involves 12 independent functions ($R$ and $1-T$ for
+each of six wavelengths). Fortunately, there is a workaround: the
+problem can be reformulated as a differentiable problem by introducing
+a dummy varaiable $t$ (the
+[epigraph](https://en.wikipedia.org/wiki/Epigraph_(mathematics))) and
+adding each independent function as a new nonlinear constraint. See
+the [NLopt
+documentation](https://nlopt.readthedocs.io/en/latest/NLopt_Introduction/#equivalent-formulations-of-optimization-problems)
+for an overview of this approach.
+
+In this example, we use a minimum feature size of 150 nm for the line
+width *and* line spacing. The implementation of these constraints is
+based on [A.M. Hammond et al., Optics Express, Vol. 29, pp. 23916-38
+(2021)](https://doi.org/10.1364/OE.431188).
+
+There are five important items to note in the set up of the
+optimization problem:
+
+- The lengthscale constraint is activated only in the final epoch. The
```

Comment:
> The initial design of the final epoch must be binary.

This isn't actually true. For this particular example, it seemed to help. But in the 80+ examples from our paper, *none* of the designs were binary when the lengthscale constraint was activated.

In fact, on feature of the lengthscale constraint is it forces the design to be binary without increasing $\beta$ to a point that breaks differentiability.

Reply:
> In fact, on feature of the lengthscale constraint is it forces the design to be binary without increasing β to a point that breaks differentiability.

In my experience, that is exactly the issue: the lengthscale constraint may change the structure too much by binarizing it too much, and the performance is kind of ruined. That's why I would personally wait until the structure is binarized enough before imposing lengthscale constraint.

Reply:
I have updated the text with additional details based on the points raised in this discussion.